### PR TITLE
Add option to impose a Dirichlet condition for p in RT example

### PR DESCRIPTION
**Pull request by lindsayad**
Error results for tris to be clear we're keeping the desired convergence rates:

grid = 4
L2 error is: 0.500032
HDiv semi-norm error is: 1.27137
HDiv error is: 1.36617
grid = 8
L2 error is: 0.251364
HDiv semi-norm error is: 0.643384
HDiv error is: 0.690743
grid = 16
L2 error is: 0.125855
HDiv semi-norm error is: 0.322662
HDiv error is: 0.346338
grid = 32
L2 error is: 0.0629499
HDiv semi-norm error is: 0.16145
HDiv error is: 0.173289
grid = 64
L2 error is: 0.0314776
HDiv semi-norm error is: 0.0807404
HDiv error is: 0.0866594

### CODE REVIEW

**Comment by lindsayad** [comment]
New BC treatment:
- 93 iterations with ILU
- condition number: 2e1

Old BC treatment:
- 148 iterations with ILU
- condition number: 6e11

**Comment by roystgnr** [comment]
I do like getting rid of penalty BCs ... but looking at the diff now, isn't this solving a different problem?  Before we were solving the PDE with Neumann BCs (on p, so they looked like Dirichlet BCs on u); now we're solving it with Dirichlet BCs on p.

**Comment by lindsayad** [comment]
Previously we were weakly enforcing the condition:

<p, n * v> = <penalty * n * (u - u_exact), n * v>

Now we are weakly enforcing the condition:

<p, n * v> = <p_exact, n * v>

where p is the scalar solution field, v is the vector test function, and n is the normals.

We are weakly enforcing a condition based on the same term that falls out of integration by parts in the vector equation

**Comment by lindsayad** [comment]
You can say that we're solving a different problem in the sense that we are involving p_exact instead of u_exact, but as we have an analytical relation between p_exact and u_exact, I don't really see them as very different. I believe that users in general will be more likely to know boundary information about the scalar solution field than the vector solution field so it makes sense to offer an example demonstrating such a case. Moreover, as shown by the diff, it's a simpler problem to implement. I can add another example if that is preferred

**Comment by roystgnr** [comment]
> we have an analytical relation between p_exact and u_exact

Do you mean "we have an exact solution that we're comparing to, so we know both p_exact and u_exact from that", or do you mean "boundary conditions on p and boundary conditions on u are equivalent problems"?  Because the former isn't true for non-manufactured problems, and the latter just isn't true.

> as shown by the diff, it's a simpler problem to implement

This is certainly true, but that's because it's different problem.  The pure Neumann boundary-value problem residual has a non-trivial kernel, so we had to have a Lagrange multiplier to constrain it away; the pure Dirichlet boundary-value problem residual doesn't, so we don't.

> I believe that users in general will be more likely to know boundary information about the scalar solution field than the vector solution field

Users are likely to know boundaries' Dirichlet data in some cases, Neumann data in others, Robin data in others, and mixtures of all three in the general (even just the general linear!) case.

> so it makes sense to offer an example demonstrating such a case.

I definitely like this.  But when we were talking on Slack I misunderstood; it sounded like the proposed changes were "solve the same problem, just with something better than a penalty method", which seemed like a pure improvement.  "Solve an easier problem instead", less so.  We're adding one bit of code with didactic value and removing two others this way.

> I can add another example if that is preferred

It'd be my preference.  My first instinct is to just put both types of boundary conditions in this example, with Dirichlet on one or two sides and Neumann on the others.  But *any* Dirichlet boundary, even one point, makes the kernel trivial and the Lagrange multiplier pointless ... whereas I really liked the Lagrange multiplier example here.  I think this was the only pure-Neumann example we had.

**Comment by lindsayad** [comment]
> Do you mean "we have an exact solution that we're comparing to, so we know both p_exact and u_exact from that", or do you mean "boundary conditions on p and boundary conditions on u are equivalent problems"? Because the former isn't true for non-manufactured problems, and the latter just isn't true.

I definitely meant the former. Yes you are right, the former is not true for non-manufactured problems, but this is a manufactured problem which is why I said it.

> We're adding one bit of code with didactic value and removing two others this way.

Yea, maybe we can just leave this example as-is. Or I can add a boolean flag to toggle between the two within this single example

**Comment by roystgnr** [comment]
> Or I can add a boolean flag to toggle between the two within this single example

Oh, I'd love that.

**Review by nmnobre — CHANGES_REQUESTED** [review]
Hi both, sorry for the delayed response.

I agree with everything that has been said, I'd also prefer to see an addition to the example instead of a simple replacement.
I'd say illustrating two alternative formulations producing the same results is actually of enormous didactic value.
Specially in this case, where both BCs are (or look like) Dirichlet, just on different fields.

Since you seem to be happy to add a boolean flag to switch between formulations, I'll mark this as "Request changes". :)

**Review by nmnobre — COMMENTED** [review]
Hi Alex, this is very nice, thank you! :)

Could you please update `doc/html/examples/vector_fe_ex6.html`? I think that's the only thing that absolutely needs doing, the other stuff is just me nit-picking... You can also add yourself as an author of this example. :)

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 1]
```diff
@@ -529,32 +544,53 @@ void assemble_divgrad(EquationSystems & es,
...
+                      // variable at the boundary, i.e. the RT vector boundary dof.
+                      const Real penalty = 1.e10;
+
+                      // A double loop to integrate the normal component of the
+                      // vector test functions (i) against the normal component of
+                      // the vector trial functions (j).
+                      for (unsigned int i = 0; i != vector_n_dofs; i++)
+                        for (unsigned int j = 0; j != vector_n_dofs; j++)
+                          {
+                            Ke(i, j) += JxW_face[qp] * penalty * vector_phi_face[i][qp] *
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:566]: Not quite sure what you mean. For instance the `RealGradient vector_value;` line in the Neumann branch is indented to column 22 and the `Real scalar_value;` line in the Dirichlet branch is also indented to column 22

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 2]
```diff
@@ -529,32 +544,53 @@ void assemble_divgrad(EquationSystems & es,
...
+                                        normals[qp] * vector_phi_face[j][qp] * normals[qp];
+                          }
+
+                      // Loop to integrate the normal component of the vector test
+                      // functions (i) against the normal component of the
+                      // exact solution for the vector variable.
+                      for (unsigned int i = 0; i != vector_n_dofs; i++)
+                        {
+                          Fe(i) += JxW_face[qp] * penalty * vector_phi_face[i][qp] * normals[qp] *
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:575]: For me the `Fe(i) +=` lines in both Neumann and Dirichlet branches are indented to column 26. The `Ke` line in the Neumann branch, due to its nested `for`, is indented to column 28

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 3]
```diff
@@ -529,32 +544,53 @@ void assemble_divgrad(EquationSystems & es,
...
+                                   vector_value * normals[qp];
+                        }
+                    }
+                  else
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:579]: I don't have a `dirichlet` boolean. In `main` we error if the user has selected something other than neumann or dirichlet so as of now there are only two possible states, so the states are coverable by the single `neumann` boolean

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (lindsayad) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:574]: Oh I understand what you're saying now... the spacing between operators

**Comment by lindsayad** [comment]
> You can also add yourself as an author

Haha nah my contributions are relatively small. I think the author line is good as-is!

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (roystgnr) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:578]: Yeah; if this was a real application we'd be testing bcids and putting Dirichlet on some and Neumann on others, but this is good for a simple example.

**Review by roystgnr — APPROVED** [review]

**Comment by roystgnr** [comment]
I kicked the Test MOOSE debug recipe ... but I think that's the second time in a week I've seen a timeout there?  That worries me.

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 4]
```diff
@@ -529,32 +544,53 @@ void assemble_divgrad(EquationSystems & es,
...
                     {
-                      Fe(i) += JxW_face[qp]*penalty*vector_phi_face[i][qp]*normals[qp]*vector_value*normals[qp];
+                      // The boundary value for scalar field.
+                      Real scalar_value;
+                      if (dim == 2)
+                        scalar_value = DivGradExactSolution().scalar(xf, yf);
+                      else
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 5]
```diff
@@ -447,51 +459,54 @@ void assemble_divgrad(EquationSystems & es,
...
-              for (unsigned int n = 0; n != lambda_n_dofs; n++)
-                {
-                  Ke(k + vector_n_dofs, n + vector_n_dofs + scalar_n_dofs) += JxW[qp]*scalar_phi[k][qp];
-                }
-
-            // A double loop to integrate the Lagrange dof (m) against the
-            // scalar trial functions (l).
-            for (unsigned int m = 0; m != lambda_n_dofs; m++)
-              for (unsigned int l = 0; l != scalar_n_dofs; l++)
+          if (neumann)
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 6]
```diff
@@ -149,8 +156,9 @@ int main (int argc, char ** argv)
   system.add_variable("u", FIRST, RAVIART_THOMAS);
   system.add_variable("p", CONSTANT, MONOMIAL);
 
-  // Add a scalar Lagrange multiplier to enforce our constraint
-  system.add_variable("l", FIRST, SCALAR);
+  // Add a scalar Lagrange multiplier to remove the nullspace if imposing the Neumann condition
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅
Haha, and from a Portuguese 😝

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 7]
```diff
@@ -238,6 +246,9 @@ void assemble_divgrad(EquationSystems & es,
   // the proper system.
   libmesh_assert_equal_to (system_name, "DivGrad");
 
+  // Retrieve our boundary condition type. If not Neumann, then it is Dirichlet
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C [hunk 8]
```diff
@@ -447,51 +459,54 @@ void assemble_divgrad(EquationSystems & es,
...
 
       // We have now reached the end of the quadrature point loop, so
       // the interior element integration has
       // been completed.  However, we have not yet addressed
-      // boundary conditions.  For this example we will only
-      // consider simple Dirichlet boundary conditions.
+      // boundary conditions.  For this Poisson example, we consider either
+      // Dirichlet or Neumann for the scalar solution field p. Note, that in
+      // the mixed formulation a Neumann condition for p corresponds to a
+      // Dirichlet condition for u
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C]: ✅

File: examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmnobre) [reply examples/vector_fe/vector_fe_ex6/vector_fe_ex6.C:578]: > I don't have a dirichlet boolean.

I know, I was trying to convince you to (even though it's obviously redundant). :)
Semantically the code is correct, and my suggestion wasn't meant as a fix, rather for readability, I was just nitpicking...

**Review by nmnobre — APPROVED** [review]
Small as they might be, I still think these changes are very valuable, so thank you! :)

**Pull request merged**